### ISSUE TITLE
Attributes of relationships

### ISSUE THREAD

**Issue by hhware** [json-api/json-api#415]
IMHO, the wording on how relationships can be specified via linkage objects is not quite clear. Are the listed attributes (`type`, `id`, `data`) the only ones allowed in specifying the relationships, or custom attributes of relationships can be listed as well?

An example of a relationship with attributes: suppose that we have a tree of directories and files in a Unix filesystem, and that we need to describe two kinds of relationships between them: direct descendance (file or subdirectory belonging to a directory) and symlink. Does the spec allow to do it by including a custom attribute directly with the linkage object? I believe that `meta` is unsuitable for this because it describes relationship as a whole and not its specific members.

Specifically, is the example below compliant?

``` http
{
  "type": "directories",
  "id": "12",
  "name": "usr",
  "links": {
    "descendants": {
      "data": [{
        "type": "directories",
        "id": "34",
        "kind": "direct"
      },{
        "type": "directories",
        "id": "56",
        "kind": "symlink"
      },{
        "type": "files",
        "id": "78",
        "kind": "symlink"
      }]
    }
  }
}
```

To allow this explicitly, I propose to change the wording in the specification of linkage as follows (proposed addition is italicized):
- Linkage to other resource objects ("object linkage") included in a compound
  document. This allows a client to link together all of the resource objects
  included in a compound document without having to `GET` one of the
  relationship URLs. Linkage **MUST** be expressed as:
  - `type` and `id` members for to-one relationships. `type` is not required
    if the value of `id` is `null`.
  - `type` and `id` members for homogeneous to-many relationships. `type` is
    not required if the value of `id` is an empty array (`[]`).
  - A `data` member whose value is an array of objects each containing `type`
    and `id` members for heterogeneous to-many relationships.
  
  _In all three cases, other members can be  OPTIONALLY included  alongside `type` and `id`
  to describe custom attributes of a relationship._
- A `"meta"` member that contains non-standard meta-information about the
  relationship.

Also, I think that the "complete example document with multiple included relationships" should be extended to include the illustration of `data` linkage.

Any chance to still consider this before the spec hits 1.0? 
I would be glad to promptly provide a PR for both proposed changes.

**Comment by hhware** [comment]
I am sorry for being inaccurate in expressing my initial suggestion, it is only applicable to the third option in the list, namely to to-many relationships (obviously, in case of RDBMS, it is only possible for the relationship to have attributes if it is stored in a separate table). Since homogeneous to-many relationships can be also represented using the third option, it makes sense to extend only the third option with custom attributes. My example above is still correct since it only covers that one case. So the updated proposal is to:

(1) Change the wording of linkage as follows (changes shown in italic):
- Linkage to other resource objects ("object linkage") included in a compound
  document. This allows a client to link together all of the resource objects
  included in a compound document without having to `GET` one of the
  relationship URLs. Linkage **MUST** be expressed as:
  - `type` and `id` members for to-one relationships. `type` is not required
    if the value of `id` is `null`.
  - `type` and `id` members for homogeneous to-many relationships. `type` is
    not required if the value of `id` is an empty array (`[]`).
  - A `data` member whose value is an array of objects each containing `type`
    and `id` members for _homogeneous or_ heterogeneous to-many relationships. 
    _Other members can be OPTIONALLY included  alongside `type` and `id` 
    to describe custom attributes of a relationship._
- A `"meta"` member that contains non-standard meta-information about the
  relationship.

(2) Update the "complete example document with multiple included relationships" to illustrate the linkage of the third kind, via `data` member.

**Comment by hhware** [comment]
(3) A correction would also be required to the section Updating a Resource's To-Many Relationships (changes are shown in italic):

If a to-many relationship is included in the `links` section of a resource
object, it **MUST** be an object containing:
- `type` and `id` members for homogeneous to-many relationships; to clear the
  relationship, set the `id` member to `[]`
- a `data` member whose value is an array of objects each containing `type` and
  `id` members _(and OPTIONALLY other custom members)_ 
  for _homogeneous or_ heterogeneous to-many relationships; to clear the
  relationship, set the `data` member to `[]`

**Comment by hhware** [comment]
[A submission](https://github.com/json-api/json-api/pull/421) of @bintoro made me reconsider my proposal once again. Sorry for bringing yet another version. OPTIONAL was the wrong choice of the term, and the scope of the problem was not adequately described.

Generally, especially since RC2, it can be said that the spec operates with two primary types of data: resource objects and relationship objects (whereas meta, errors, etc. can be considered auxiliary data). Out of these two primary types of data, resource objects can have arbitrary attributes, but the spec is silent about the ability of the relationship objects to have attributes. To me, this means that the spec allows them in extensions, but mainstream implementations are unlikely to support them. However, to-many relationships are represented by separate tables in case of RDBMS backend, and therefore have full potential to have attributes of their own. It is not clear (at least to me) how to represent them in the framework of JSON API.

IMHO, supporting relationship attributes could be useful in many practical scenarios, especially in cases when an API is built around an edge-labeled multigraph (e.g., when edges have associated weights/costs) or a hierarchical content tree where linking of subrees with other subtrees is allowed. 

Another, much more widespread use case could be the ability to store audit data, such as creation time and creator ID (which is a to-one relationship itself) as attributes of to-many relationships. It can be argued that relationships with attributes can be simply represented via resource objects, and the problem goes away. However, if we imagine an abstract system, about which we only know that it requires (a) soft deletion and (b) storage of some minimal audit data (creator ID / creation time / deleter ID / deletion time) for all its objects and relationships, all relationships in such a system will have to be represented as resource objects in JSON API, which is quite unnatural. I believe that JSON API could be made more flexible to accommodate such generic use case.

This issue can be thought as not blocking 1.0. However, the larger point here is whether relationships are just nameless edges connecting nodes, or can they be allowed to have metadata. IMHO, this larger point is intrusive enough and practical enough to be considered before tagging 1.0. 

If at all possible, I would like to have a chance to prepare a PR to present this proposal in a more clear form. I will definitely be able to do it by this Sunday, March 8.

**Comment by bintoro** [comment]
~~@hhware, can you help me understand what you're trying to accomplish here? PR #421 fully supports attributes on relationships, you just have to prefix them. Is it the prefix that is a problem?~~ Okay I think I get it now... sorta.

**Comment by dgeb** [comment]
@hhware resources of different types can in fact be returned in a single collection. Your `kind` qualifier _could_ be contained in the `type` field (not saying that's the right answer though).

> However, if we imagine an abstract system, about which we only know that it requires (a) soft deletion and (b) storage of some minimal audit data (creator ID / creation time / deleter ID / deletion time) for all its objects and relationships, all relationships in such a system will have to be represented as resource objects in JSON API, which is quite unnatural.

There are a few possibilities:
- The audit data should not leak out into the API
- All this metadata makes a case for these "relationships" being resources, complete with links to the `creator`, `deleter`, etc. If it needs to be accessible by clients, then it needs to be in a normal form.

If you still want to try to make a case for this and you can't do it until Sunday, March 8, please do so by 5pm GMT. I'm not sure exactly when we're going to try to tag 1.0, and I'm not sure where you're located, but I hope this seems reasonable. 

I'll tag this 1.0 so we don't forget.

**Comment by hhware** [comment]
@bintoro No, prefix is not a problem at all. Please correct me if I am wrong, but I think that the position expressed by the PR #421 in the context of the current JSON API spec is the following: 
- resource objects can have arbitrary attributes (in addition to several specific required ones), and the mainstream library implementations must support these arbitrary attributes (support means be aware of their existence and provide facilities to handle them),
- all other objects, including relationship objects, must have specific attributes, and the major library implementations must support these specific attributes,
- any object can be extended with any custom attributes (their names can start with non-alphanumeric characters for all objects, although this restrictions is lifted for meta and error objects), and the major implementations do not have to support these custom attributes.

My thinking was the following:
- resource objects and relationship objects can have arbitrary attributes (in addition to several specific required ones), and the mainstream library implementations must support these arbitrary attributes (motivation: in case of RDBMS backend, resources and to-many relationships can be represented by separate tables and can therefore naturally have attributes),
- all other objects must have specific attributes, and the major library implementations must support these specific attributes,
- any object can be extended with any custom attributes (same naming rules as above), and the major implementations do not have to support these custom attributes.

@dgeb Thanks for the opportunity! 5 pm GMT on March 8 is totally reasonable.

> resources of different types can in fact be returned in a single collection. Your kind qualifier could be contained in the type field (not saying that's the right answer though).

I agree, it can work pretty well if the custom attribute is represented with a small set of strings. It seems to be quite a stretch if that set of strings is large, or is there are multiple custom attributes per relationship. It seems impossible if the custom attribute is numeric. 

> The audit data should not leak out into the API

It is a reasonable limitation, I accept it.

> All this metadata makes a case for these "relationships" being resources, complete with links to the creator, deleter, etc. If it needs to be accessible by clients, then it needs to be in a normal form.

I agree, this option is always available. However, I think that often such modeling would look unnatural, in which case it could be viewed more like a limitation of JSON API, which is fine, as long is it not an omission.

**Comment by hhware** [comment]
@steveklabnik @tkellen It was my impression from the discussion in #431 that this issue is postponed until after 1.0 because the change proposed here is additive. Indeed, after #421, arbitrary attributes are allowed only in resource objects and in `meta`, but this list could be extended in 1.x to include relationship objects, if the change proposed here is accepted in some form or shape. Is there still any interest in continuing this discussion (e.g., reviewing mockups which I was going to provide), or closing of this issue should be considered its final resolution?

**Comment by steveklabnik** [comment]
Sorry, it was automatically closed by GitHub because the commit wasn't re-worded.

**Comment by hhware** [comment]
@dgeb, @tkellen, @steveklabnik, what would be your opinion on allowing relationship attributes via official extension? The extension will be purely additive, the use case is clear (the fact that #431 was a ["tough call"](https://github.com/json-api/json-api/pull/431#issuecomment-92938203) appears to confirm that), semantics of proposed change has a direct match in the semantics of RDBMSes, the new conteinerized structure of resources and relationships will make the difference between resources and relationships with attrubutes visible at the first glance. Given all that, I doubt that presence of such extension could possibly hurt the spec. Moreover, it might be useful to have it as a probe to measure utility of this feature, and the rate of its adoption would allow to evaluate how practical was your judgement in the ["struggle to balance complexity and utility"](https://github.com/json-api/json-api/pull/431#issuecomment-92938203).

Since the PR on this would be essentially a rewrite of #431, i.e. the full scope of it is already known, I hope I could get a definitive answer on the _acceptance of the concept_ before I start.

**Comment by tkellen** [comment]
cc @bobholt if he has time to weigh in on his recent encounter writing an API specification for a client of ours that will likely wind up being non-compliant with JSON-API due to the decision to not allow this.

**Comment by tkellen** [comment]
@hhware We will likely be setting up a registry of some kind. It's my hope that being "official" won't mean much. I'd like to empower the community to freely create extensions that are easily discoverable, and to slowly roll the most popular additive ones into the base specification over time.

Long answer short, I am very interested in seeing a revision to this concept as long as it is additive to the base specification.

**Comment by hhware** [comment]
> Long answer short, I am very interested in seeing a revision to this concept as long as it is additive to the base specification.

Thanks! I will write another PR :)

**Comment by tkellen** [comment]
Awesome!
